### PR TITLE
Public goods

### PR DESCRIPTION
**Pull request by Schwartz10**
Adds a Wrapper proxy that deposits a donated portion of iFIL to the owner of the proxy - used for public goods funding

### CODE REVIEW

File: src/PublicGoodsDonator.sol [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+    if (donationPercentage > DENOM) {
+      revert InvalidParams();
+    }
+    _;
+  }
+
+  IPreStake public immutable preStake;
+
+  IWFIL private wFIL;
+  IERC20 private iFIL;
```
Comment (PinqKiwi) [inline src/PublicGoodsDonator.sol]: I would change this name - it's not directly related to FIL and it looks like an interface with the way the name is now which is misleading.

File: src/PublicGoodsDonator.sol [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    iFIL = _poolToken;
+  }
+
+  /**
+   * @notice Deposit WFIL into the PreStake contract
+   * @param recipient The account to pull WFIL from and forward the iFIL tokens to
+   * @param amount The amount of WFIL stake
+   * @param donationPercent The percentage of the amount to split to the PG wallet. 100 = 100%
+   */
+  function deposit(
```
Comment (PinqKiwi) [inline src/PublicGoodsDonator.sol:57]: The assumption here is that the ``donationPercent`` is the percent of ``amount`` that this contract will _retain_ and the rest will get set to the recipient, correct?

File: src/PublicGoodsDonator.sol [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+  /// @dev modifier to ensure `donationPercentage` is 100% max
+  modifier noOverDonations(uint256 donationPercentage) {
+    if (donationPercentage > DENOM) {
+      revert InvalidParams();
+    }
+    _;
+  }
+
+  IPreStake public immutable preStake;
```
Comment (PinqKiwi) [inline src/PublicGoodsDonator.sol:33]: Might want to be able to set/upgrade this - probably not necessary

**Review by PinqKiwi — APPROVED** [review]
LGTM, I don't love the name ``iFIL`` just cause it implies it's FIL based (which it isn't necessarily) and also it looks like an interface. Otherwise I don't see any issues.

I removed some unused imports and events, there was pause/unpause events in here and just to be clear there's no pause or anything for this contract - I assume we're basically relying on the ``mint`` in ``PreStake`` to fail out if the minter has been changed, I think this is a good solution just confirming.

Might be good to do some ``expectReverts`` against the ``Owner``.

File: src/PublicGoodsDonator.sol (hunk 1 above)
Comment (Schwartz10) [reply src/PublicGoodsDonator.sol]: See your point for the code, tho this literally _is_ `iFIL`

File: src/PublicGoodsDonator.sol (hunk 2 above)
Comment (Schwartz10) [reply src/PublicGoodsDonator.sol:57]: Correct

File: src/PublicGoodsDonator.sol (hunk 3 above)
Comment (Schwartz10) [reply src/PublicGoodsDonator.sol:33]: Would rather actually just use pause/unpause on it

**Pull request merged**